### ISSUE TITLE
[HOLD for payment 2023-09-27] [$2000] Chat - Main composer is displayed when editing two messages

### ISSUE THREAD

**Issue by techievivek** [Expensify/App#23908]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

### Case 1:

1. Send two messages (message A and message B)
2. Switch to edit mode in message A
3. Switch to edit mode in message B
4. Hide keyboard (press back button)
5. Main Composer is shown but it shouldn't because focus is still on message B

### Case 2:

1. Send two messages (message A and message B)
2. Switch to edit mode in message A
3. Switch to edit mode in message B
4. Click the cancel button of message A
5. Main Composer is shown but it shouldn't because focus is still on message B

### Case 3:

1. Send two messages (message A and message B)
2. Switch to edit mode in message A
3. Open EmojiPicker
4. Close EmojiPicker
5. Switch to edit mode in message B
6. Click the cancel button of message A
7. Main Composer is shown but it shouldn't because focus is still on message B

**Notes:**
- For case 1 we may have a fix for that already (the fix didn't introduce a regression) ([ref1](https://github.com/Expensify/App/pull/23618#issuecomment-1652702584), [ref2](https://github.com/Expensify/App/issues/17531#issuecomment-1646614420)) 
- For cases 2 and 3 we need to make sure we are not re-introducing the regression caught in [#24016](https://github.com/Expensify/App/issues/24016)

Additional testing (regression) steps:

1. Send a message (message A)
2. Switch to edit mode in message A
3. Clear the edit field
4. Press save
5. Confirm deletion prompt
6. Verify Main Composer is shown because there is no other focused message

## Expected Result:
Main compose box should not be displayed

## Actual Result:
Main compose box is displayed

## Workaround:
Don't edit two messages at once.

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.47

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/232604613-64146c4c-9b6c-47c3-bbe3-839122ff6006.mp4

https://github.com/Expensify/App/assets/35863227/00f9d5e3-6bbf-44f3-a668-6d6dff8a1c83

**Expensify/Expensify Issue URL:**

**Issue reported by:** @s77rt 

**Slack conversation:** https://expensify.slack.com/archives/C02NK2DQWUX/p1690561889854179

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01230125fc5ec18c7d</li>
        <li>Upwork Job ID: 1685886732158468096</li>
        <li>Last Price Increase: 2023-08-30</li>
<li>Automatic offers: </li>
<ul>
<li>s77rt | Reviewer | 26608183</li>
<li>s-alves10 | Contributor | 26608186</li>
<li>s77rt | Reporter | 26608188</li>
</ul></li>
    </ul>
</details>

**Comment by techievivek** [comment]
For context, we discovered and discussed about this issue here: https://expensify.slack.com/archives/C02NK2DQWUX/p1690561889854179

**Comment by techievivek** [comment]
@s77rt Can you confirm if this issue is reproducible on other platforms? or just on Android?

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Main composer is displayed when editing two messages when the first one is cancelled if we open Emoji picker and close

### What is the root cause of that problem?

When we cancel or save the draft, the following code runs
https://github.com/Expensify/App/blob/024d210bc0b5a0454b62a72de60057af67cdd045/src/pages/home/report/ReportActionItemMessageEdit.js#L123-L129

We consider the case when `isFocusedRef.current` is true. `EmojiPickerAction.isActiveReportAction` checks the last active report action
https://github.com/Expensify/App/blob/024d210bc0b5a0454b62a72de60057af67cdd045/src/components/EmojiPicker/EmojiPicker.js#L114

Even though we hide the emoji picker, last active report action still remains and this function returns true. So the if condition returns false and `ComposerActions.setShouldShowComposeInput(true);` is called

This is the root cause

### What changes do you think we should make in order to solve the problem?

We can solve this issue by setting last report action to empty when we hide the emoji picker

Update the `hideEmojiPicker` method of [EmojiPicker] as follows
```
    const hideEmojiPicker = (isNavigating) => {
        if (isNavigating) {
            onModalHide.current = () => {};
        }
        emojiPopoverAnchor.current = null;
        setIsEmojiPickerVisible(false);
        setReportAction({});
    };
```

This works as expected

<details>
    <summary>Result</summary>


</details>

### What alternative solutions did you explore? (Optional)

**Comment by s77rt** [comment]
@techievivek Yes, on all platforms (as long as the window width is small). Just a note the second video in OP does not reflect the current state. https://github.com/Expensify/App/pull/23618 is not deployed to staging yet.

**Comment by s77rt** [comment]
@s-alves10 Thanks for the proposal. Your RCA is correct. The suggested solution although it looks like it would work it may cause a race condition [^1] [^2]. The alternative solution looks more like a workaround (we are not clearing the report action in the right place).

It would better to fix the first solution race condition e.g. have `ReportActionItemMessageEdit` `useEffect` run before `ReportActionItem` or find another solution to https://github.com/Expensify/App/issues/22214.

[^1]: https://github.com/Expensify/App/pull/23618#issuecomment-1654058421
[^2]: https://github.com/Expensify/App/pull/23618#issuecomment-1654776446

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

**Comment by s77rt** [comment]
@dukenv0307 Thanks for the proposal. I don't think your RCA is correct. This issue only occurs after opening the emoji picker. Can you please update your proposal taking this into account?

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The Composer shows up even though the focus is still in message B. This happens when another message A had been edited before and had also used the Emoji Picker.  

### What is the root cause of that problem?
As already mentioned, the report action remains active even though the emoji picker is hidden. This causes the composer to show up.

### What changes do you think we should make in order to solve the problem?

To solve the problem, whenever the `ReportActionItemMessageEdit` is unmounted, we can safely hide the EmojiPicker if this is an active report action. We can replace [this code](https://github.com/Expensify/App/blob/6a383413758030d687a728f66b19b00eadab7f3c/src/pages/home/report/ReportActionItemMessageEdit.js#L123C1-L127C14) as follows.
This will fix the current issue and #22214.

```
            if (EmojiPickerAction.isActiveReportAction(props.action.reportActionID)) {
                EmojiPickerAction.hideEmojiPicker(true);
            }

            // Skip if this is not the focused message so the other edit composer stays focused.
            if (!isFocusedRef.current) {
                return;
            }
```

### What alternative solutions did you explore? (Optional)

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The Composer shows up even though the focus is still in message B. This happens when another message A had been edited before and had also used the Emoji Picker.  

### What is the root cause of that problem?
As already mentioned, the report action remains active even though the emoji picker is hidden. This causes the composer to show up.

### What changes do you think we should make in order to solve the problem?

To solve the problem, whenever the `ReportActionItemMessageEdit` is unmounted, we can safely hide the EmojiPicker if this is an active report action. We can replace [this code](https://github.com/Expensify/App/blob/6a383413758030d687a728f66b19b00eadab7f3c/src/pages/home/report/ReportActionItemMessageEdit.js#L123C1-L127C14) as follows.
This will fix the current issue and #22214.

```
            if (EmojiPickerAction.isActiveReportAction(props.action.reportActionID)) {
                EmojiPickerAction.hideEmojiPicker(true);
            }

            // Skip if this is not the focused message so the other edit composer stays focused.
            if (!isFocusedRef.current) {
                return;
            }
```

### What alternative solutions did you explore? (Optional)

**Comment by s77rt** [comment]
@rojiphil Thanks for the proposal but isn't this proposed already by @s-alves10?

**Comment by rojiphil** [comment]
> @rojiphil Thanks for the proposal but isn't this proposed already by @s-alves10?

But, he is also suggesting to add `isEmojiActive`. If that is required, then, his proposal is a good choice. I do not think that is required. That way, my proposal is different.

**Comment by techievivek** [comment]
Hey, I have updated the issue description to include the original bug where the main composer gets focused when editing two messages.

👋 @garrettmknight coming from this bug https://github.com/Expensify/App/issues/17531#issuecomment-1662191253. This issue is linked to https://github.com/Expensify/App/issues/17531, which we decided to close due to the PR being reverted. Since the original issue is complex and was priced at $2000, I'll update this one to start with the same price.

**Comment by techievivek** [comment]
@s77rt 
> @techievivek Yes, on all platforms (as long as the window width is small). Just a note the second video in OP does not reflect the current state. https://github.com/Expensify/App/pull/23618 is not deployed to staging yet.

I see, in that case, can you please ensure the proposals include fix for both emoji picker and the race condition bug. Also, let me know if you want me to update anything in the OP, thanks.

**Comment by s-alves10** [comment]
@s77rt 

Do we need to update the issue description? I think it should include https://github.com/Expensify/App/issues/17531 as well

**Comment by s77rt** [comment]
Yeah, we have 4 cases (at least) to handle:
- 2 mentioned here https://github.com/Expensify/App/issues/17531#issuecomment-1646614420
- This one about the emoji picker
- The forgotten one (regression) about saving an empty message https://github.com/Expensify/App/issues/24016

Will have to summarise those before updating the OP, but feel free to post proposals

**Comment by rojiphil** [comment]
> Yeah, we have 4 cases (at least) to handle:

@s77rt   Thanks for the feedback and details. There is quite some history to this issue which I was unaware of. Will look at this more closely and see if I am able to propose something here.

**Comment by s77rt** [comment]
@techievivek Please update the description to:

## Action Performed:

### Case 1:

1. Send two messages (message A and message B)
2. Switch to edit mode in message A
3. Switch to edit mode in message B
4. Hide keyboard (press back button)
5. Main Composer is shown but it shouldn't because focus is still on message B

### Case 2:

1. Send two messages (message A and message B)
2. Switch to edit mode in message A
3. Switch to edit mode in message B
4. Click the cancel button of message A
5. Main Composer is shown but it shouldn't because focus is still on message B

### Case 3:

1. Send two messages (message A and message B)
2. Switch to edit mode in message A
3. Open EmojiPicker
4. Close EmojiPicker
5. Switch to edit mode in message B
6. Click the cancel button of message A
7. Main Composer is shown but it shouldn't because focus is still on message B

## Expected Result:
Main compose box should not be displayed

## Actual Result:
Main compose box is displayed

**Comment by s77rt** [comment]
**Notes:**
- For case 1 we may have a fix for that already (the fix didn't introduce a regression) ([ref1](https://github.com/Expensify/App/pull/23618#issuecomment-1652702584), [ref2](https://github.com/Expensify/App/issues/17531#issuecomment-1646614420)) 
- For cases 2 and 3 we need to make sure we are not re-introducing the regression caught in [#24016](https://github.com/Expensify/App/issues/24016)

Additional testing (regression) steps:

1. Send a message (message A)
2. Switch to edit mode in message A
3. Clear the edit field
4. Press save
5. Confirm deletion prompt
6. Main Composer is not shown but it should because there is no other focused message

**Comment by techievivek** [comment]
@s77rt Updated, thanks.

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

@s77rt 
I made a slight changes considering the various cases. It works fine on web platform. I'll test on all platforms soon

**Comment by s77rt** [comment]
@s-alves10 Thanks for the update. I haven't read the proposal yet but at first glace looks like there is a lot of code :sweat_smile: It would be great if you can summarise the solution of each case in a sentence or two e.g. https://github.com/Expensify/App/issues/17531#issuecomment-1646614420. Regardless I will review asap. I'm also looking forward for platforms test.

**Comment by s77rt** [comment]
@s-alves10 Any update on the above?

**Comment by s-alves10** [comment]
Will update soon

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

**Comment by s77rt** [comment]
@s-alves10 Thanks for the update.

## Case 1
I think it's clear as we'd be reapplying the previous solution.

## Case 2
### Root cause
When we delete or unmount a message, we show the composer no questions asked

### Solution
We should show the main composer only if we delete or unmount the **active** message. A message is considered active if:
1. It's focused: This is a straightforward check as we have a state that tracks the focus :heavy_check_mark: 
2. **It has context menu open:** Let's discuss this point. We can check if the report action has context menu open using `ReportActionContextMenu.isActiveReportAction` but as I understand this does not work as expected because once we delete the message we [clear the report action](https://github.com/Expensify/App/blob/0dcb69907051c403a09fed0a56080e3441a8a5eb/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L331) (too early) and when we check `isActiveReportAction` it returns false, we deleted an active message but lost the info that tells us that - this will make composer not show when needed (regression). Can you explain how we are going to handle this scenario?
3. It has emoji picker open: Hold 

## Case 3
Hold

**Comment by s-alves10** [comment]
@s77rt 
#### Case 2
I don't suggest to clear the report action on unmount or delete. I am suggesting to clear the report action when other report action gets focused.

**Comment by s77rt** [comment]
@s-alves10 I'm not saying that we should clear it. I'm saying it's already being cleared [here](https://github.com/Expensify/App/blob/0dcb69907051c403a09fed0a56080e3441a8a5eb/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L331) right? So calling `isActiveReportAction` in `ReportActionItemMessageEdit`'s `useEffect` may be too late.

**Comment by s-alves10** [comment]
@s77rt 
In my proposal, I suggested to manage confirm delete modal and context menu separately. We need to check report action for confirm delete modal when checking if we should show composer

**Comment by s-alves10** [comment]
@s77rt 
I was wrong. I think the above code should be deleted. We need to clear the reportActionID on focus event

**Comment by s-alves10** [comment]
@s77rt 

I've tested all 4 cases mentioned above and works fine. Race condition won't be happen.
I'll find more edge cases. 
By the way, can you tell me your thoughts on my solution?

**Comment by s77rt** [comment]
@s-alves10 I would really prefer not to add another state management for context menu, there should be only one, the confirm modal is supposed to be a part of the context menu.

> I think the above code should be deleted. We need to clear the reportActionID on focus event

I think it's more logical to clear the report action when the context menu is hidden. There can be cases where we just hide the context menu and not focus on another report action.

**Comment by s-alves10** [comment]
@s77rt 

Here are my thoughts.
> I would really prefer not to add another state management for context menu, there should be only one, the confirm modal is supposed to be a part of the context menu.

We have some interactions related to the report action in edit mode. They are emoji picker and delete modal.
We don't have context menu for actions in edit mode. So I wanted to separate context menu and delete modal, and to handle emoji picker and delete modal similarly. In addition, I think there is a big gap between context menu and delete modal. Why do they need to be handled as one state?

> I think it's more logical to clear the report action when the context menu is hidden. There can be cases where we just hide the context menu and not focus on another report action.

I thought handling emoji picker and delete modal similarly would be good. For emoji picker, we have race condition if we clear the report action when it hides. So I suggest not to clear it on hide. I think similar race condition can be happened for delete modal as well.

These are just my humble opinion. Expecting your feedback

**Comment by s77rt** [comment]
@s-alves10 The delete modal is currently a part of `ReportActionContextMenu` they should have same state for the time being. I think the problem with that is timing right? (we need to make sure events go in a certain order) If this approach is getting too complicated it may be a good idea to take a step back and find another approach.

(Let's not worry about the emoji picker for now)

**Comment by rojiphil** [comment]
> 2\. **It has context menu open:** Let's discuss this point.

@s77rt I was checking the various scenarios here but do not understand the use case when context menu is open. Are the steps to reproduce this mentioned somewhere? Please let me know. Thanks. Sorry if I missed this somewhere.

**Comment by rojiphil** [comment]
> @rojiphil Edit a message, clear the input then press the save button. This will show the delete modal which is a part of the context menu.

Got it. Thanks. Investigating...

**Comment by s-alves10** [comment]
@s77rt 

I think we can only one state for `ReportActionContextMenu` as you mentioned.

We can solve the case 2 as follows
We should show the main composer only if we delete or unmount the active message. A message is considered active if:

1. It's focused: This is a straightforward check as we have a state that tracks the focus
2. It has context menu open. We can check if the report action has context menu open using `ReportActionContextMenu.isActiveReportAction`.
Don't clear the `reportAction` for `ReportActionContextMenu` in `modalHide`. We clear it when an action gets focused. And we set `reportAction` when context menu is opened or delete modal is opened without clearing it. This way `isActiveReportAction` returns true when on `unmount` or `deleteDraft` call

**Comment by s77rt** [comment]
@s-alves10 Thanks for the update. For the most part I like the idea and the consistency how we handle both the emoji picker and context menu in the same way. Unfortunately we have a problem:

> Don't clear the reportAction for ReportActionContextMenu in modalHide.

We can't just undo that because this was added here https://github.com/Expensify/App/pull/20185 to fix https://github.com/Expensify/App/issues/18165. We need to fix the original issue in another way before we can remove that code.

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There are focus issues when multiple messages are been edited. Also, there are consistency issues with the show/hide of the Composer when multiple messages are edited. The scope is covered in the mentioned use cases in the problem description which needs to be addressed.

### What is the root cause of that problem?

*Note: Message Editor refers to the editor of a message which is a report action.*

The current implementation hides the Composer when the [Message Editor gets focus](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L326) and shows the Composer whenever the Message Editor is [blurred](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L340) or [draft is deleted](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L205). The Composer is also shown on [unmount](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L131) of the Message Editor if it is the currently focused one as [shown here](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L125). The current approach relies on showing/hiding the Composer when the focus comes in and goes out of the Message Editor. 

Hiding the Composer when Message Editor gets focused is the right thing to do. Then, the next logical question to ask is 'When do we show the Composer?' We are currently trying to handle this during the [unmount here](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L122-L132) and [on blur here](https://github.com/Expensify/App/blob/ab99bf4748d03e309257e24a8828b48a118ba5b9/src/pages/home/report/ReportActionItemMessageEdit.js#L340). There, we are trying to identify all the cases where we should skip showing the Composer. Now, that would be a challenge as we have to target all individual cases. Further, showing/hiding the Composer will mount/unmount itself and will also bring/lose focus on the Composer. When there are multiple Message Editors and the focus needs to be on one of the Message Editors, this will bring a competition between Composer and Message Editor to gain focus. According to me, these are the problems that we need to address to resolve this problem.  

### What changes do you think we should make in order to solve the problem?

This proposal is an alternate approach.
The proposed solution is to move the decision of where to put focus at a later point when Composer is mounted or props get updated. Further, instead of identifying all individual cases where we should skip showing the Composer, let us show the composer only when it is needed. 

The implementation can be as follows:

1) We know that there can be only editor (i.e. Composer or Message Editor) which can gain focus. To ensure that the appropriate editor gets focus, let us introduce a callback `focusActionItemCallback` for Message Editor just like we have `focusCallback` [here](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/libs/ReportActionComposeFocusManager.js#L5) for Composer. Further, we can update `onComposerFocus`, `focus` and `clear` to accommodate `focusActionItemCallback` in ReportActionComposeFocusManager.js. Also, we can add a helper function `isCurrentActionItem` to identify if the Message Editor is the currently focused one. 

2) To use `focusActionItemCallback`, let us first call `onComposerFocus` for the Message Editor when it [gains focus here](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/pages/home/report/ReportActionItemMessageEdit.js#L324). Additionally, we can clear the callback for Message Editor whenever a) Composer [gains focus here](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/pages/home/report/ReportActionCompose.js#L347) and b) Message Editor is unmounted (addressed in point-4). Now, whenever `focus` gets called, Composer or Message Editor gets focused depending on the current use case.

3) Now, let us call `focus` at the required places. First, when the Composer is mounted, we can replace [this.focus() here](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/pages/home/report/ReportActionCompose.js#L740) with `ReportActionComposeFocusManager.focus()`. Also, the same can be done for [update here](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/pages/home/report/ReportActionCompose.js#L277). By making the decision on where to focus at the time of mount/update of Composer, we can avoid competition between Composer and Message Editor for gaining focus. Also, we can remove [these lines](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/pages/home/report/ReportActionItemMessageEdit.js#L205-L206) as `deleteDraft` will always unmount the Message Editor and we will handle this case in point-4 below.

4) As mentioned earlier, let us show the composer only when it is needed. Now, we know that the only case when we need to show the composer is if the currently focused Message Editor is been unmounted. We can leverage the recently added helper function `isCurrentActionItem` for this purpose. Also, when the Message Editor unmounts, we need to bring the focus back to the right place (i.e. Composer or another Message Editor). We can do this by replacing [this code](https://github.com/Expensify/App/blob/e587008e1b985810be7e5bdc07e256564c6d8215/src/pages/home/report/ReportActionItemMessageEdit.js#L122C1-L133C1) with the following:

```
    useEffect(() => {
        return () => {
            if (ReportActionComposeFocusManager.isCurrentActionItem(onFocusCallback))
            {
                ReportActionComposeFocusManager.clear(true);
                ComposerActions.setShouldShowComposeInput(true);
            }
            ReportActionComposeFocusManager.focus();
        };

    }, []);
```

5) Case 2 and Case 3 in the problem statement should get resolved with the above implementation. Regarding Case 1, the [keyboardDidHide handler](https://github.com/Expensify/App/blob/48824f9904399ae5719d0af5ccb0e93f66f116b2/src/libs/openReportActionComposeViewWhenClosingMessageEdit/index.native.js#L5) triggers the problem. For me, Case 1 gets fixed by doing just like the [web version here](https://github.com/Expensify/App/blob/ea78897933ca70e687f52d24564192e6d0e3011f/src/libs/openReportActionComposeViewWhenClosingMessageEdit/index.js#L4). I am not clear why the `keyboardDidHide` handler was added in the first place. Or may be I am missing some information here due to internal discussions. @s77rt , would you know the issue that will get reproduced if we do not use the `keyboardDidHide` handler?

### What alternative solutions did you explore? (Optional)

**Comment by rojiphil** [comment]
@s77rt I have tested the [proposal](https://github.com/Expensify/App/issues/23908#issuecomment-1673883823) across platforms and it seems to work well. Please do share your thoughts on the proposal. Also, let me know if there are any additional cases that you would like me to test. Thanks.

**Comment by s77rt** [comment]
@rojiphil Thanks for the proposal. If I understood the concept, we will replace the "focused" state management: instead of using the `isFocused()` method we will be using our own. We will set the flag as focused in the native focus callback but we won't unset it in the native blur callback instead we will unsent it when we focus on another input or delete that input. Would that be about right?

Also the focus callback or `focusActionItemCallback` is not needed here, right?

**Comment by rojiphil** [comment]
> If I understood the concept, we will replace the "focused" state management: instead of using the `isFocused()` method we will be using our own. We will set the flag as focused in the native focus callback but we won't unset it in the native blur callback instead we will unsent it when we focus on another input or delete that input. Would that be about right?
> 
> Also the focus callback or `focusActionItemCallback` is not needed here, right?

@s77rt You are right that we will replace the "focused" state management by using our own. Further, as you mentioned, we will not unset the flag in the native blur callback but we will unset it when we focus on another input or delete that input. However, there is an additional point to take care of. A deleted input may or may not be the currently focused one. We need to know this because we show the Composer only for a deleted input that was currently in focus. A generic flag will not help here. This is where an input-specific `focusActionItemCallback` will help. We compare `focusActionItemCallback` with the deleted input's focus callback to determine if this is the one which was in focus (covered in [point-4 of the proposal](https://github.com/Expensify/App/issues/23908#:~:text=As%20mentioned%20earlier)). So, I think, `focusActionItemCallback` is needed. And, by the way, `focusActionItemCallback` is also used as a flag here.

**Comment by s77rt** [comment]
@rojiphil Can you please clarify what would we exactly store in `focusActionItemCallback`?

**Comment by rojiphil** [comment]
@s77rt `focusActionItemCallback` can store a reference to the following callback within `ReportActionItemMessageEdit` component:

```
    const onFocusCallback = useRef(
        () => {
            textInputRef.current.focus();
        },
        [textInputRef],
    );   
```
    
Hope this clarifies. Please let me know if any more information is required.

**Comment by s77rt** [comment]
@rojiphil Can we compare two refs? or what would we compare in `isCurrentActionItem`?

**Comment by rojiphil** [comment]
@s77rt Since useRef returns a reference to Object, `focusActionItemCallback` and `onFocusCallback` can be simply compared with `===` as follows:

```
function isCurrentActionItem(onFocusCallback) {
    return focusActionItemCallback === onFocusCallback;
}
```
I don't know if there is a better solution for this. Would be glad to explore if any.

**Comment by s77rt** [comment]
@rojiphil Got it. I think that's a valid approach but it may get too complex if we want to cover all the cases. For instance I think if we open the emoji picker of any message, the `isCurrentActionItem` would return `false` because we didn't focus on the input. This would mean if we delete that message the Main Composer won't show up.

**Comment by s77rt** [comment]
@rojiphil Sorry I meant if that message is deleted from another device while emoji picker is still open from your side

**Comment by rojiphil** [comment]
> I meant if that message is deleted from another device while emoji picker is still open from your side

@s77rt Got it. I tried this scenario too. Note the deletion of messages `Hello One` and `Hello Two` from another device. LGTM. The test video is on the device on my side. 

https://github.com/Expensify/App/assets/3004659/5bea8c52-f276-4b03-a32f-3a595d828e69

**Comment by rojiphil** [comment]
> but it may get too complex if we want to cover all the cases.

@s77rt, tbh, I don't think our concept is complex. Actually, we have taken a step or two back and have taken a holistic view to consistently address the show/hide of Composer and manage the focus on multiple message edits and Composer. But, I agree that number of cases is many as it has been already evident through the history of issues related to this topic.

**Comment by s-alves10** [comment]
@s77rt 

Regarding https://github.com/Expensify/App/issues/23908#issuecomment-1673669023, I suggest to clear active report action below
https://github.com/Expensify/App/blob/21d483c098621b59feb0522475d75ceb13e4e2a8/src/pages/home/ReportScreen.js#L174

```
        ReportActionContextMenu.clearActiveReportAction();
```

I think this is reasonable in the sense that it reset the context menu when open a report. Let me know your thoughts

**Comment by rojiphil** [comment]
@s77rt Here are a few more test videos with the iOS simulator. Updates from Onyx due to remote delete of inputs will trigger either unmount or blur events locally on inputs, and depending on the use case the Main Composer is shown. 

<details>
<summary>Remote Deletion Test 1</summary>

https://github.com/Expensify/App/assets/3004659/62801b51-88be-417e-bde8-adc3ad8f0ba4

</details>

<details>
<summary>Remote Deletion Test 2</summary>

https://github.com/Expensify/App/assets/3004659/7cda2b27-a7d2-41c5-b1ac-2502c7267cfd

</details>

I can also share a test version on Git Hub if that helps in reviewing the concept in more detail. Please let me know.

**Comment by rojiphil** [comment]
@techievivek  @s77rt A humble request. If possible, can we increase the bounty for this issue as the number of surrounding cases to take care of is more than the listed ones in this issue (e.g. deleting from the remote device)? And, especially, if we are taking a step back and taking a holistic view as [mentioned here](https://github.com/Expensify/App/issues/23908#issuecomment-1677910068), I think there is even a stronger case for the increase in bounty. What are your thoughts?

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

**Comment by s77rt** [comment]
@s-alves10 Thanks for the update. I think the alternative solution for https://github.com/Expensify/App/issues/18165 is a workaround, changing the report have nothing to do with the active context menu.

**Comment by s-alves10** [comment]
@s77rt 

Yeah. That's not a solution. It'll cause a regression

In order to solve https://github.com/Expensify/App/issues/24016, I think removing the clear action in `modalDidHide` is unavoidable. It would be good if we can clear the active report action when the report action is unmounted, but this isn't the solution for https://github.com/Expensify/App/issues/18165 because we use the stack navigator and unmounting doesn't happen. So we need another solution for doing this.

I suggest the creation of isContextMenuVisible() function. We have `isPopoverVisible` state already. We highlight the report action only when hovered or context menu is visible. I think this is reasonable and tested well.

Let me know your thoughts

**Comment by rojiphil** [comment]
> When we trigger the unmount, we will check if `isCurrentActionItem` but that's supposed to be false because the input was never focused, so the main composer won't show up.
> For the case you demonstrated the composer is shown because it was already visible, it seems to show up when you press the emoji picker button so the bug is not reproducible but I guess this is only the case for native, on mWeb I think the bug may be actually reproducible.

@s77rt Thanks for the feedback. LGTM on mWeb too. Attaching test videos for the same. Agree that `isCurrentActionItem` will not be true in the demonstrated case and Main Composer will not be shown due to this. However, the `ReportActionComposeFocusManager.focus()` [added here](https://github.com/Expensify/App/issues/23908#:~:text=ReportActionComposeFocusManager.focus()%3B) will try to focus on the input but will be blurred due to the presence of emojipicker/context-menu. This ensures that the Main Composer is shown. This is consistent behavior across native/mWeb.

<details>
<summary>Emoji Picker Test </summary>

https://github.com/Expensify/App/assets/3004659/2402c011-4caa-42cb-b8ea-1d3452c93ca0

</details>

<details>
<summary> Context Menu Test</summary>

https://github.com/Expensify/App/assets/3004659/6203ff90-be44-41cf-b962-78818b1224e0

</details>

**Comment by s77rt** [comment]
@rojiphil Why do we have the `ReportActionComposeFocusManager.focus()` outside the if condition?

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

@s77rt 

`isContextMenuVisible` isn't needed. We need `isDeleteConfirmModalVisible` function. I updated the proposal accordingly

**Comment by rojiphil** [comment]
> @rojiphil Why do we have the `ReportActionComposeFocusManager.focus()` outside the if condition?

@s77rt That's a good question. We added this outside the if condition because where the focus should be is independent of the message input being unmounted. Fundamentally, we keep our two objectives (i.e. show/hide of Main Composer and the focus management) separate and address them.

**Comment by s77rt** [comment]
@s-alves10 Thanks for the update. I think `isDeleteConfirmModalVisible` is still not needed. `isActiveReportAction` is the only function that should be used. Otherwise we would be just adding another condition to identify when a message is active (the 3 mentioned [here](https://github.com/Expensify/App/issues/23908#issuecomment-1668704630) and this "It has delete confirm modal visible".

**Comment by s77rt** [comment]
@rojiphil It's still not clear, why we are trying to focus the Main Composer when any of the messages is deleted? Seems like a workaround.

**Comment by rojiphil** [comment]
@s77rt Sorry for not being very clear with my explanation. Let me explain. As mentioned in [point 1 in proposal here](https://github.com/Expensify/App/issues/23908#:~:text=Further%2C%20we%20can%20update%20onComposerFocus%2C%20focus%20and%20clear%20to%20accommodate%20focusActionItemCallback%20in%20ReportActionComposeFocusManager.js), `focusActionItemCallback` is also integrated into `ReportActionComposeFocusManager.focus()` to have centralized focus management. Further, as mentioned in [point 4 of proposal here](https://github.com/Expensify/App/issues/23908#:~:text=Also%2C%20when%20the%20Message%20Editor%20unmounts%2C%20we%20need%20to%20bring%20the%20focus%20back%20to%20the%20right%20place), we call `ReportActionComposeFocusManager.focus()` to ensure that focus remains with another focused message input when an explicit mouse click has resulted in the unmount of this unfocused message input. This is the reason why `ReportActionComposeFocusManager.focus()` is outside the if condition. Hope my explanation is better now. Please let me know.

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

@s77rt 
You're right. I've found a solution. We can change the confirm callback of the `showDeleteModal`. Please check the updated proposal

**Comment by s77rt** [comment]
@rojiphil That's still looking as a workaround. There is no real reason to call `ReportActionComposeFocusManager.focus()` when any message is deleted.

Any alternative solution for fixing this without introducing a new focus management?

**Comment by s77rt** [comment]
@s-alves10 I don't see what changes we are making to `showDeleteModal`. We were already using the `isFocusedRef` that way here https://github.com/Expensify/App/pull/23618. Also the `isFocusedRef` is available without the need to pass it as a parameter.

**Comment by s-alves10** [comment]
@s77rt 

Yes. That's the point. The value of `isFocusedRef.current` is different when we click save button and we click delete button. RIght? I wanted to use its value at the time we click save button.

**Comment by rojiphil** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1673883823)

@s77rt I get your point. Have updated the proposal to remove the new focus management and instead use `isActiveReportAction`. Please refer the section of Alternative Solutions for the details. Thanks.

**Comment by getusha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Main composer displayed editing two messages
### What is the root cause of that problem?
When `ReportActionItemMessageEdit.js` is unmounted and deleteDraft is called, we will set setShouldShowComposeInput to true. This action makes the main composer visible regardless of the status of other editable composers.

### What changes do you think we should make in order to solve the problem?

Instead of setting the value of `shouldShowComposeInput` we can add a counter above and When we set the value to 'false', we'll increase the counter. And when we set it to 'true', we'll decrease the counter. We'll only set shouldShowComposeInput to 'true' when the counter is at 0. This way, we'll make sure that the compose input is shown only when no other ReportActionItem is being edited.

To avoid the unreliability of using a number counter we can track how many active ReportActionitems are being edited by making the counter an array and push the unique index of the ReportActionItem that is being edited to it. using the length of the array we will check if it is empty or not, empty means there is no item being edited.

For example, let's say we have a list of `editItems` with numbers `[index1, index2]`. This means we're currently working on editing two items in ReportAction. When we delete a draft, we'll remove it from the list using its matching index. this will make sure, the primary composer will remain hidden until any ongoing edits of ReportAction items are canceled.


https://github.com/Expensify/App/assets/75031127/dc162797-f165-4539-b5eb-3c0708aba64e


### What alternative solutions did you explore? (Optional)
N/A

cc @s77rt

**Comment by s-alves10** [comment]
@s77rt 

In order to use the value at the time we click save button, we use this type of call: `() => deleteDraft(isFocusedRef.current)`. If we don't use parameter and use the `isFocusedRef.current` inside the `deleteDraft` function, this is not the case. But we use `pass by value` here in this syntax.

**Comment by s77rt** [comment]
@getusha Thanks for the proposal. Your RCA is correct. I think the solution may not work if we open a report that already has some report actions in edit mode as the counter at this point would be zero.

**Comment by getusha** [comment]
@s77rt that's a valid case, here after implementing the solution.

https://github.com/Expensify/App/assets/75031127/be2122e5-d3b4-460d-947d-1df504ae886a

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

Here is the demonstration code
```
  const isFocusedRef = { current: false };
  const callback = (() => { const focused = isFocusedRef.current; return () => console.log(focused); })()
  setTimeout(callback, 1000);
  isFocusedRef.current = true;
```

**Comment by s-alves10** [comment]
@getusha 

If there is no focused draft, the main composer should be shown, I think

**Comment by getusha** [comment]
@s-alves10 my solution is also flexible to handle that case.

https://github.com/Expensify/App/assets/75031127/bc08a58d-a4ca-4a63-bcd2-822932761e76

**Comment by s77rt** [comment]
@s-alves10 Oh I got it! Indeed that's a good solution and we can have it without that weird function wrapper, we can just pass the `isFocused` (from state and not from the ref): `() => deleteDraft(isFocused)`. However unfortunately this won't work if the message is deleted from another device while you have the delete modal shown up because the logic in the `useEffect` is still not taking the context menu in consideration.

**Comment by s-alves10** [comment]
@s77rt 

Yeah. `&& !ReportActionContextMenu.isActiveReportAction()` should be added to the `if` condition, I think

**Comment by rojiphil** [comment]
> this solution won't work as intended because `ReportActionContextMenu.isActiveReportAction` won't be true as you'd except and this is due to the changes made by this PR #20185.

@s77rt
If I understand you correctly, your concern is that since `reportAction` is reset [here during onHide](https://github.com/Expensify/App/blob/dd2c98b839b0aa85f4ffa3bc6f230cf1d499183d/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L331C25-L331C25), `ReportActionContextMenu.isActiveReportAction` will not return true. But, I am confident that it will return true. The reason is that ContextMenu and DeleteModal are hidden when the parent component `ReportActionItem` is destroyed as shown [here](https://github.com/Expensify/App/blob/08dd014c77a32782415612015daa7efd065bf200/src/pages/home/report/ReportActionItem.js#L139C1-L142C14). And, `ReportActionItemMessageEdit` is a child component of `ReportActionItem`. As such, the unmount of `ReportActionItemMessageEdit` will be called first. At this time,  the `reportAction` will not be lost. 

As mentioned in my proposal, the only thing that is required is to show the Main Composer if there are EmojiPicker/ContextMenu shown for the deleted message input. No need to check for any other flags or even hide the emojipicker/contextmenu. Please let me know your thoughts.

**Comment by s77rt** [comment]
@getusha I think it's better to avoid counters here. There may be more report actions in edit mode than you can see. Also I don't think counters will fix case 2 outlined above.

**Comment by s77rt** [comment]
@s-alves10 Any alternative solution for https://github.com/Expensify/App/issues/18165? That should probably help us a ton

**Comment by getusha** [comment]
@s77rt is case 2 the following?

>Case 2:
Send two messages (message A and message B)
Switch to edit mode in message A
Switch to edit mode in message B
Click the cancel button of message A
Main Composer is shown but it shouldn't because focus is still on message B

**Comment by s-alves10** [comment]
@s77rt 

I've found a better solution. We can clear it in confirm/cancel callback here
https://github.com/Expensify/App/blob/402c70c7b5fe5b0f7ee7bb656d112d0e003ebe0e/src/pages/home/report/ContextMenu/ContextMenuActions.js#L306-L315

There are only two calls: above call and the call in `ReportActionItemMessageEdit`

**Comment by s77rt** [comment]
@rojiphil 
> If I understand you correctly, your concern is that since reportAction is reset [here during onHide](https://github.com/Expensify/App/blob/dd2c98b839b0aa85f4ffa3bc6f230cf1d499183d/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L331C25-L331C25), ReportActionContextMenu.isActiveReportAction will not return true

Yes, that's correct.

> The reason is that ContextMenu and DeleteModal are hidden when the parent component ReportActionItem is destroyed as shown [here](https://github.com/Expensify/App/blob/08dd014c77a32782415612015daa7efd065bf200/src/pages/home/report/ReportActionItem.js#L139C1-L142C14)

I don't think that's relevant, the modal is also hidden if you just press the confirm button.

> As such, the unmount of ReportActionItemMessageEdit will be called first

Interestingly this is not true if the message is deleted from another device. But as mentioned in the above point, this may actually be unrelated.

**Comment by s77rt** [comment]
@getusha Yes

**Comment by s-alves10** [comment]
@s77rt 

In the call in `ReportActionItemMessageEdit`, we don't clear report action in callbacks.

**Comment by rojiphil** [comment]
@s77rt 
Further to my [comment here](https://github.com/Expensify/App/issues/23908#issuecomment-1684253862) and your [reply here](https://github.com/Expensify/App/issues/23908#issuecomment-1684284211). My comments were specifically for the case of deletion of message input from another device. I also tested this in the test video on mWeb as shown below. Isn't this the test case? Or am I missing something here?  Please let me know.

https://github.com/Expensify/App/assets/3004659/53a4955e-75c0-4fd0-8f88-29703b3d8f83

**Comment by s77rt** [comment]
@s-alves10 It's not clear, can you clarify when to clear the report action exactly?

**Comment by s-alves10** [comment]
> the unmount of ReportActionItemMessageEdit will be called first. At this time, the reportAction will not be lost.

@rojiphil Are you sure?

**Comment by rojiphil** [comment]
> Are you sure?

@s77rt @s-alves10 Have just sent a test video [here](https://github.com/Expensify/App/issues/23908#issuecomment-1684292336). Isn't this the test case? If not, share with me the specific test case and I will verify it again.

**Comment by s-alves10** [comment]
@s77rt 

We set empty functions as confirm/cancel callbacks in `showDeleteModal` call in context menu now. We can set confirm/cancel callback there like `() =>ReportActionContextMenu.clearActiveReportAction()`. Context menus doesn't work for messages in edit mode.

Change nothing in `showDeleteModal` call in `ReportActionItemMessageEdit`
Clear active report action in focus event as I proposed earlier

**Comment by s77rt** [comment]
@rojiphil https://github.com/Expensify/App/issues/24016

**Comment by getusha** [comment]
>Main Composer is shown but it shouldn't because focus is still on message B

@s77rt will the focus remain on message B? Could you please provide me with a video? can be from staging or prod.

**Comment by s-alves10** [comment]
This issue has a lot of edge cases. Hoping all contributors read all history of this issue and https://github.com/Expensify/App/issues/17531, https://github.com/Expensify/App/issues/24016

**Comment by s77rt** [comment]
@s-alves10 

> We can set confirm/cancel callback there like () =>ReportActionContextMenu.clearActiveReportAction().

Is that meant to be a solution for https://github.com/Expensify/App/issues/18165?


> Clear active report action in focus event in ReportActionItemMessageEdit

We should clear the report action when the context menu is hidden as that's the natural way and not based on another focus event

**Comment by rojiphil** [comment]
> @rojiphil #24016

@s77rt Thanks for sharing that. The issue you mentioned above is exactly the one I tested and sent the [video here](https://github.com/Expensify/App/issues/23908#issuecomment-1684292336). Just curious, why do you think this should not work?

**Comment by s-alves10** [comment]
@s77rt 

> Is that meant to be a solution for #18165?

yes

> We should clear the report action when the context menu is hidden as that's the natural way

What did you expect from https://github.com/Expensify/App/issues/23908#issuecomment-1684261955?

In addition, if we clear report action in modal hide, race condition can appear as in `EmojiPicker`, I think

**Comment by rojiphil** [comment]
> @rojiphil #24016

@s77rt Extremely sorry. Here, you were not referring to remote delete but local delete. Let me check this.

**Comment by getusha** [comment]
@s77rt incase you missed my comment https://github.com/Expensify/App/issues/23908#issuecomment-1684305387

**Comment by s77rt** [comment]
@getusha Yes the focus should remain on message B. You can apply this PR https://github.com/Expensify/App/pull/23618 to see the wanted results for that case.

**Comment by s77rt** [comment]
@s-alves10 
> Clear active report action in focus event in ReportActionItemMessageEdit

Do we need to do that? What if we kept the value untouched same as done in the emoji picker? (we don't change the report action in focus event for the emoji picker right?)

**Comment by rojiphil** [comment]
@s77rt 
> Interestingly this is not true if the message is deleted from another device.

The test video [here ](https://github.com/Expensify/App/issues/23908#issuecomment-1684292336) demonstrates that it works the way I expected i.e. unmount of `ReportActionItemMessageEdit` first, then, `ReportActionItem` unmount is called.

> I don't think that's relevant, the modal is also hidden if you just press the confirm button.

If we press the confirm button, the focus is already on the message input. So, the logic in unmount of focused message input will ensure that the composer is shown.

**Comment by s77rt** [comment]
@rojiphil 

> unmount of ReportActionItemMessageEdit first, then, ReportActionItem unmount is called.

I tested this with console.log and `ReportActionItem` is actually the first. But regardless we need to fix the case where we delete the message locally as well.

**Comment by getusha** [comment]
@s77rt i see the expected result, it can be applicable to my proposal.
here is the result: 

https://github.com/Expensify/App/assets/75031127/3f903035-f88b-4792-9eb3-a4ea700cf3c5

**Comment by s77rt** [comment]
@getusha Can we please avoid using a counter? As mentioned earlier this may cause an issue when dealing with old report actions that are left in edit mode.

**Comment by getusha** [comment]
@s77rt which means we only want to handle this case for Report actions which are most recent?

**Comment by rojiphil** [comment]
> I tested this with console.log and `ReportActionItem` is actually the first. But regardless we need to fix the case where we delete the message locally as well.

@s77rt You are right. `ReportActionItem` is actually the first. Sorry about that. 

Here are my findings. The reason why ContextMenu will work in my proposal is this:
In the case of ContextMenu, whether it is `Remote Deletion` or `Local Deletion`, the `focusActionItemCallback` will refer to this message input. So, the unmount logic of `ReportActionItemMessageEdit` will kick in to show the Main Composer as mentioned in [point(4) of my proposal](https://github.com/Expensify/App/issues/23908#issuecomment-1673883823)

Does this make sense now?

**Comment by s77rt** [comment]
@getusha 

> which means we only want to handle this case for Report actions which are most recent?

No, the solution should not relay on other report actions at all.

> i would love if you could explain what issue exactly causes

1. Send two messages (message A and message B)
2. Switch to edit mode in message A (counter = 1)
3. Switch to edit mode in message B (counter = 2)
4. Delete message B from another device (counter = 1)
5. The Main Composer should show up but it won't because the counter is not zero

**Comment by s-alves10** [comment]
@s77rt 

> Do we need to do that? What if we kept the value untouched same as done in the emoji picker? (we don't change the report action in focus event for the emoji picker right?)

We need to clear active action for emoji picker as well. We told that before

**Comment by s77rt** [comment]
@rojiphil Are you referring to `ReportActionComposeFocusManager.focus();`? That's a failsafe call. I don't think it's really needed (outside the if condition). It's only there as a workaround - we shouldn't fall into that fail scenario.

**Comment by s77rt** [comment]
@s-alves10 Ah for case 3, right sorry about that.

> Instead of clear active action in modal hide, we can clear it in confirm/cancel callback here

How is that different? `onModalHide` we will still call `callbackWhenDeleteModalHide` which will call our callback.

**Comment by s-alves10** [comment]
@s77rt 

Yes. So if we don't clear active action in the callback, active action won't be cleared. right?
We need to clear active action for context menu's delete modal. And we shouldn't clear active action for delete modal in publishDraft of ReportActionItemMessageEdit

**Comment by s77rt** [comment]
@s-alves10 I'm not sure what we are trying to solve here? This is supposed to be another fix for https://github.com/Expensify/App/issues/18165 (and we revert the linked PR) But that solution is pretty much the same and won't do us any good as `ReportActionContextMenu.isActiveReportAction` would still return `false`.

**Comment by getusha** [comment]
>Send two messages (message A and message B)
Switch to edit mode in message A (counter = 1)
Switch to edit mode in message B (counter = 2)
Delete message B from another device (counter = 1)
The Main Composer should show up but it won't because the counter is not zero

@s77rt if you see the places where we call `ComposerActions.setShouldShowComposeInput(true);` you can see it's also called when the component is unmounted right? which includes when the report item is deleted. please let me know if there are other test cases i should test against.


https://github.com/Expensify/App/assets/75031127/94b517ec-6e7c-48c5-91df-a7fccac05896

**Comment by s77rt** [comment]
@getusha In your attached video the main composer did show before we get to switch to the other message. Try to focus again on the message.

1. Focus on message A (counter = 1)
2. Focus on message B (counter = 2)
3. Delete message B from another device (counter = 1)
4. The Main Composer should show up but it won't because the counter is not zero

**Comment by getusha** [comment]
@s77rt when the focus is removed the counter will be decreased you are aware of that right?

**Comment by s77rt** [comment]
@getusha It's not the case when the focus is transferred to another message

**Comment by getusha** [comment]
>@getusha It's not the case when the focus is transferred to another message

@s77rt  In order to focus Message B, message A will be blurred eventually, when that happens the counter will decrease.


https://github.com/Expensify/App/assets/75031127/5b0cee6f-cccf-4bf3-ba35-9c3fce0bfc9f


let me know if i did something wrong while testing.

**Comment by s77rt** [comment]
@getusha I'm not sure how this is working. It does not go with your proposal. Why is the counter in this case zero (or why it's working)?

> message A will be blurred eventually, when that happens the counter will decrease

According to [`onBlur` logic](https://github.com/Expensify/App/blob/03cf0b14f82a41db004dfee27799e3cea1608ee5/src/pages/home/report/ReportActionItemMessageEdit.js#L373-L376) we will only call `openReportActionComposeViewWhenClosingMessageEdit` if the next focused element is not another message input, thus we wouldn't decrease the counter in this case.

**Comment by s-alves10** [comment]
@s77rt 

>  I'm not sure what we are trying to solve here? This is supposed to be another fix for #18165 (and we revert the linked PR) But that solution is pretty much the same and won't do us any good as ReportActionContextMenu.isActiveReportAction would still return false.

What I want to do here is to remove the clearing action in `modalDidHide`. What is important is that we can control whether clear or not the active action. If we don't clear the active action in confirm/cancel callback, `ReportActionContextMenu.isActiveReportAction` would return true.

**Comment by getusha** [comment]
> @getusha I'm not sure how this is working. It does not go with your proposal. Why is the counter in this case zero (or why it's working)?

@s77rt I can share the diff where i call the functions if you are interested 😄

**Comment by rojiphil** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1673883823)

@s77rt

> Are you referring to `ReportActionComposeFocusManager.focus();`?

I am not referring to `ReportActionComposeFocusManager.focus();` for solving the ContextMenu issue.

Please do have a look at the updated Alternative Solutions section in the proposal. I have attempted to bring more clarity to the changes required. 
Also, please let me know your thoughts. 

_Extract from the Proposal_
**Alternative Solution to new focus Management:**
**Purpose**: 
a) Remove the usage of `ReportActionComposeFocusManager.focus()` from the main proposal as there is no real reason to call this. [Reference comment here](https://github.com/Expensify/App/issues/23908#issuecomment-1683093060)
b) Solve the edge case in mWeb i.e. Main Composer does not show up when EMojiPicker is already displayed for an unfocused message input and the same message input is deleted from the remote device. [Reference comment here](https://github.com/Expensify/App/issues/23908#issuecomment-1678938912)
**Note**: Usage of `isCurrentActionItem` is still very much there as mentioned in the main proposal.

**Comment by s77rt** [comment]
@s-alves10
> If we don't clear the active action in confirm/cancel callback, ReportActionContextMenu.isActiveReportAction would return true

Right, but we have to clear it, otherwise we would reintroduce https://github.com/Expensify/App/issues/18165 or is there a condition to decide whether to clear it or not?

**Comment by s77rt** [comment]
@getusha 
> I can share the diff

Code diffs are discouraged. Can you please explain the raised concern in plain English?

**Comment by s77rt** [comment]
@rojiphil Can we please avoid adding a new focus management. I think we already have the info we need, we just need to manage them well. We have the focus info about the input itself, its emoji picker and its context menu.

**Comment by s-alves10** [comment]
@s77rt 

As I said before, we clear that in context menu delete. This will prevent https://github.com/Expensify/App/issues/18165
And we don't clear that in `publishDraft`. This makes the `ReportActionContextMenu.isActiveReportAction` return true

**Comment by s-alves10** [comment]
@s77rt 

We need to clear the active action on unmount before showing main composer

```
            if (ReportActionContextMenu.isActiveReportAction(props.action.reportActionID)) {
                ReportActionContextMenu.clearActiveReportAction();
            }
            if (EmojiPickerAction.isActiveReportAction(props.action.reportActionID)) {
                EmojiPickerAction.clearActiveReportAction();
            }
```

That is, if a message in edit mode is an active action for ReportActionContextMenu or EmojiPickerAction, we should clear active action when the draft is unmounted

**Comment by s77rt** [comment]
@s-alves10 This still does not completely solve the regression


https://github.com/Expensify/App/assets/16493223/8d849fd6-bb9b-456d-9968-8ee5e57e7f93

**Comment by s-alves10** [comment]
@s77rt 

It works fine for me. I don't understand how the above regression happens. Will you explain the reason for the above regression in words?

**Comment by s77rt** [comment]
@s-alves10 After we clear the report action, the report action won't re-render and the highlight is still there (if you navigate away and back the highlight won't be there). Please notice the difference between the two videos I sent (those are 2 cases that are a bit different).

**Comment by getusha** [comment]
> Code diffs are discouraged. Can you please explain the raised concern in plain English?
>>According to [onBlur logic](https://github.com/Expensify/App/blob/03cf0b14f82a41db004dfee27799e3cea1608ee5/src/pages/home/report/ReportActionItemMessageEdit.js#L373-L376) we will only call openReportActionComposeViewWhenClosingMessageEdit if the next focused element is not another message input, thus we wouldn't decrease the counter in this case.

@s77rt the function to remove/decrease the counter will be called right below/above, ` setIsFocused(false)` since the checks are not relevant to the action we are performing here.

**Comment by s-alves10** [comment]
@s77rt 

Yeah. I've tested those 2 cases and works fine. 
Whenever the active report action in edit mode unmounts, we clear the active action. So when you cancel the message in edit mode, active is cleared and can't be highlighted.

**Comment by s77rt** [comment]
@getusha That would work I suppose but I still don't think using a counter is the best approach here. Ideally we won't need any new info to store.

**Comment by s77rt** [comment]
@s-alves10 Can you please test against main? If possible also include a video for the two cases, perhaps I'm missing something.

**Comment by getusha** [comment]
> @getusha That would work I suppose but I still don't think using a counter is the best approach here. Ideally we won't need any new info to store.

@s77rt May i ask the reason why we don't want to store any new info? comparing it to other proposals which are more over complicated for this issue, will using a counter bring any issues? i appreciate your input. 🙇

**Comment by s-alves10** [comment]
@s77rt 

Yes. I applied the changes from the current main branch and it did cause a regression mentioned [here](https://github.com/Expensify/App/issues/23908#issuecomment-1687155281)

I suggest to clear the report action in `modalDidHide` as it is now and use the `() => deleteDraft(isFocused)`

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat the Main composer is displayed when editing two messages

### What is the root cause of that problem?

With recent changes from this PR, the root cause might change. It's now caused by clearing the editcomposer callback here:

https://github.com/Expensify/App/blob/4f412db5e26782740529eb0ef17b5db157628413/src/pages/home/report/ReportActionItemMessageEdit.js#L215

### What changes do you think we should make in order to solve the problem?

I think we can remove that line and the original issue from https://github.com/Expensify/App/issues/23898 and cases in this issue will be resolved. We restore the cursor to last focused input instead of main composer

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat the Main composer is displayed when editing two messages

### What is the root cause of that problem?

With recent changes from [this PR](https://github.com/Expensify/App/pull/24481), the root cause might change. It's now caused by clearing the editcomposer callback here:

https://github.com/Expensify/App/blob/4f412db5e26782740529eb0ef17b5db157628413/src/pages/home/report/ReportActionItemMessageEdit.js#L215

### What changes do you think we should make in order to solve the problem?

I think we can remove that line and the original issue from https://github.com/Expensify/App/issues/23898 and cases in this issue will be resolved. We restore the cursor to last focused input instead of main composer

**Comment by s77rt** [comment]
@getusha
> May i ask the reason why we don't want to store any new info?

 I think we have enough info to determine whether the composer should be shown or not.

**Comment by s77rt** [comment]
@s-alves10 
> I suggest to clear the report action in modalDidHide as it is now and use the () => deleteDraft(isFocused)

I think we discussed this earlier. This won't work if the message is deleted from another device. We can't have the logic restricted to `deleteDraft` function, we need to cover the `useEffect` callback as well.

**Comment by s77rt** [comment]
@wildan-m Thanks for the proposal. That PR seems to have some impacts on this issue indeed. Thanks for linking! But the root cause is still not clear. Can you please elaborate?

Please note that the use of `_.defer` or `setTimeout` is strictly discouraged as they are workarounds.

**Comment by wildan-m** [comment]
@s77rt will use `InteractionManager.runAfterInteractions` is also discouraged?

**Comment by wildan-m** [comment]
@s77rt proposal updated with `focusWithDelay` option to make sure we perform focus when `ComposerFocusManager.isReadyToFocus` is true

**Comment by rojiphil** [comment]
> @rojiphil Can we please avoid adding a new focus management. I think we already have the info we need, we just need to manage them well. We have the focus info about the input itself, its emoji picker and its context menu.

@s77rt

I am not suggesting to use the new focus management. What I am simply suggesting is to replace the current `isFocusedRef` with another one (say `currentFocusedMessageEdit`). `isFocussedRef` is just an indicator of whether this message input has `focus` or is `blurred` and is not fully within our control. However, what we really need to know during `unMount` is if this message input was where the focus should have been. If we know this, I don't think we have to worry about temporary UI elements (like context menu, emoji picker, and any other UI that are introduced in the future) being displayed at the time of deletion of the draft. For example, when the focus is on message input A and its deletion occurs via Context Menu's Delete Dialog, `isFocusedRef` will return false but `currentFocusedMessageEdit` will return true. We can use this information to show the Main Composer. In this case, we do not need to think about depending on active `ReportAction` of the context menu. 

Also, if `focusActionItemCallback` is causing confusion here, we can replace this with `currentFocusedMessageEdit` which can refer to the `textInputRef` of `ReportActionItemMessageEdit`. 

The only exception I noticed in my proposal is an edge case in mWeb as [per your comment here](https://github.com/Expensify/App/issues/23908#issuecomment-1677780418). But, this can be fixed by showing Main Composer during Unmount when the following conditions are met: i.e. a) EmojiPicker is visible, b) EmojiPicker's active action is for this message input, and c) This is a mobile browser.

Managing with existing values of `isFocusedRef` and active ReportAction of emojipicker/context menu is already a challenge due to the risk of regressions. Compared to this, isn't my proposal simple and straightforward? 

Please let me know your thoughts.

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

**Comment by getusha** [comment]
>I think we have enough info to determine whether the composer should be shown or not.

@s77rt you probably are correct, i am just comparing the solution to existing proposals here, which utilize a bunch of new info including the Keyboard.

**Comment by s77rt** [comment]
@wildan-m The use of `InteractionManager.runAfterInteractions` is also discouraged as it uses `setTimeout` internally. For case 2, I think your solution is easier but does not work if we are working with 3+ messages


https://github.com/Expensify/App/assets/16493223/8273cf5a-9467-4dba-a68f-13f45e9a5521

**Comment by s77rt** [comment]
@rojiphil I understand your solution but we are essentially creating a new focus management because we will have two sources that tell us whether the input is focused or not: `isFocused` (managed natively) and `currentFocusedMessageEdit` (managed manually i.e. we populate this variable).

**Comment by s77rt** [comment]
@s-alves10 Thanks for the update. The solution for https://github.com/Expensify/App/issues/23908#issuecomment-1687155281 looks promising, but I see another problem (unrelated to this issue) which is that isContextMenuActive is a state. I don't think a state is needed here, can we just have `isContextMenuActive = ReportActionContextMenu.isActiveReportAction`?

> There is another issue mentioned [here](https://expensify.slack.com/archives/C049HHMV9SM/p1692803895784489).

This seems like the wrong url, can you please double check?

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

@s77rt 

I'm sorry. Wrong URL was updated
Regarding `isContextMenuActive`, I think we need to have this as state because dynamic update should re-render the component. For example, when [toggleContextMenuFromActiveReportAction](https://github.com/Expensify/App/blob/7fa5690dc1197266cd892195e17dea9bfd1338f3/src/pages/home/report/ReportActionItem.js#L201) is called, we need to re-render the component. Please let me know your thought

**Comment by rojiphil** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/23908#issuecomment-1673883823)

@s77rt

The proposal is updated with `Solution 1` under `What changes...`.  This solution attempts to resolve the problem by remaining within the scope of ReportActionItemMessageEdit.js. Along with `isFocusedRef`, this approach relies on states for ContextMenu and EmojiPicker which can be appropriately set/reset when context menu/Emoji Picker are shown/hidden.
More details on implementation are included within the proposal under heading `Solution 1`.
Please let me know if this is fine.

**Comment by s77rt** [comment]
@s-alves10 I think we should not worry about the state use in that function for now. Same thing goes for the other reported slack issue. Can you please just double check that all the 3 cases are fixed in both mWeb and Native and works well if message is deleted either locally or remotely.

**Comment by s77rt** [comment]
@rojiphil Thanks for the update. I don't think we need states for ContextMenu and EmojiPicker because we already have global refs for those and we have functions that return whether a given report action is active.

**Comment by techievivek** [comment]
This is a tricky one and may require some more time to come up with a proper fix(context here: https://github.com/Expensify/App/issues/23908#issuecomment-1662210868).

**Comment by garrettmknight** [comment]
@s-alves10 can you give us an update when you get a chance? Thanks.

**Comment by s-alves10** [comment]
@s77rt 

These are the test steps I've done

Note: Step 5 - 6 are for mobile platforms.
1. Open ND and log in with any account
2. Open any chat(chat C) and send two messages(message A and message B)
3. Switch message A to edit mode
4. Switch message B to edit mode
5. Hide keyboard(press back button)
6. Verify that the main composer isn't shown
7. Click the cancel button of message A
8. Verify that the main composer isn't shown
9. Switch message A to edit mode again
10. Open EmojiPicker on message A and close it
11. Click message B to focus on it
12. Click the cancel button of message A
13. Verify that the main composer isn't shown
14. Delete all contents of message B and press Enter or click Save
15. Confirm delete modal appears and click cancel
16. Verify that the main composer isn't shown
17. Again press Enter or click Save
18. Click delete on confirm delete modal
19. Verify that the main composer is shown
20. Switch message A to edit mode
21. Click Save button and click cancel on confirm delete modal
22. Switch to another chat and return back to chat C
23. Click cancel in message A
24. Verify that the message A isn't highlighted

The solution looks working fine.

I think it would be good to double the bounty. This issue has a lot of edge cases

**Comment by s77rt** [comment]
@s-alves10 

> Again press Enter or click Save
> Click delete on confirm delete modal
> Verify that the main composer is shown

Can you please explain how is this working? The main composer is not showing due to some recent PR that got merged and reintroduced https://github.com/Expensify/App/issues/24016

**Comment by s-alves10** [comment]
@s77rt 

Yes, when press Enter, the issue appears. I explained about this

The root cause of this issue: when deleting message, after the message gets focused and the unmount handler is called in a quick succession after `deleteDraft` is called. When get focused, `setIsFocused(true)` is called but `isFocusedRef.current` value doesn't change until this effect callback is called. The `unmount` handler is called just before this ref value is updated and so the handler returns here.

In order to solve this issue, we need to sync the isFocused and isFocusedRef.current. So remove the above useEffect and set isFocusedRef.current whenever setIsFocused is called

**Comment by s77rt** [comment]
@s-alves10 Any idea which PR broke this?
> The unmount handler is called just before this ref value is updated and so the handler returns here.

I don't think this is the case. The useEffect callbacks are called in order. isFocusedRef is always updated first before executing the cleanup callback.

**Comment by s-alves10** [comment]
@s77rt 

I verified this by logging

Yes, `useEffect` callback is called in order. but the cleanup function of `useEffect` callback is called before any effect callback is called.

**Comment by rojiphil** [comment]
> This is a tricky one

@techievivek @s77rt, Where do we draw the line when it comes to introducing two local states and solving the problem in a straightforward manner vis-a-vis trying to reuse global refs and trying hard to avoid regressions due to global usage of these refs? Curious to learn if I am missing something here regarding best practices followed at Expensify. For context, please compare [Solution 1 in the proposal here](https://github.com/Expensify/App/issues/23908#issuecomment-1673883823) and [Case 2 and Case 3 Solution in the proposal here](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680).

**Comment by s77rt** [comment]
@rojiphil There should be only one source of truth.

**Comment by s-alves10** [comment]
@s77rt

Please check https://legacy.reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

After the cleanup is called, effect callback isn't called because the cleanup is unmount in our case

**Comment by s77rt** [comment]
@s-alves10 How is that relevant? We are not interested in the effect of the second callback. We have the following:
- callback-1
- callback-2
- cleanup-callback-2


callback-1 is always executed before cleanup-callback-2. Thus isFocusedRef is updated correctly (in callback-1).

callback-2 is irrelevant here.

**Comment by s-alves10** [comment]
@s77rt 

You can check that isFocusedRef value in the cleanup function

**Comment by s77rt** [comment]
@s-alves10 Ah so the state is not getting updated since the update is async and we happen to unmount the component before setting the new state.

**Comment by s77rt** [comment]
@s-alves10 OK I got it it but it's not clear how this suddenly broke. Any idea?

**Comment by s-alves10** [comment]
@s77rt 

I've checked the old version(18th Jul) and there was already this issue. This is an old issue, not introduced recently

**Comment by s77rt** [comment]
@s-alves10 I don't know for sure but something is different, this was not a bug. It was introduced once here https://github.com/Expensify/App/issues/24016 then after we revered the offending PR the bug was no longer reproducible. Now it's _suddenly_ reproducible again.

**Comment by s-alves10** [comment]
@s77rt 

https://github.com/Expensify/App/issues/24016 was the issue that happens when click send button. This works fine now. The current issue happens when press Enter key. They are different

**Comment by s77rt** [comment]
@s-alves10 I have been checking this and the root cause seems related to the Main Composer showing up when pressing Enter due to the call of `openReportActionComposeViewWhenClosingMessageEdit` in `onBlur`. This function is not called when pressing the save button because it's excluded (`saveButtonID`).

**Comment by s77rt** [comment]
I think this issue is complex enough. The [newly reported issue](https://expensify.slack.com/archives/C049HHMV9SM/p1692747837798019) should be handled separately unless it's a blocker which does not seem to be the case, right? cc @s-alves10

**Comment by s-alves10** [comment]
@s77rt 

I think so, but I'm afraid the new issue would be a regression of the PR as before.

**Comment by s-alves10** [comment]
@s77rt 

Yeah. You're right. The new issue can't be a regression. I agree

**Comment by s77rt** [comment]
@s-alves10 Ok great. In this case let's proceed and fix the current reported bugs for this issue.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/23908#issuecomment-1657736680)

**Comment by bondydaa** [comment]
hmm unassigning myself since an internal engineering is already involved and i assume @techievivek has way more context here.

**Comment by garrettmknight** [comment]
@techievivek can you help move this one along?

**Comment by techievivek** [comment]
Sure, I can handle this one.

**Comment by techievivek** [comment]
Great job on the proposal, @s-alves10! And thanks, @s77rt, for helping move this forward and for handling numerous conversations. 🙇

**Comment by thienlnam** [comment]
QA was still able to reproduce this issue

**Comment by wildan-m** [comment]
@thienlnam, can you share a video?

**Comment by s-alves10** [comment]
@thienlnam 

There are 3 cases in this issue this PR solved. Which case wasn't solved?
Please share the reproduction steps

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified:
    - Case 1: https://github.com/Expensify/App/pull/15201
    - Case 2: https://github.com/Expensify/App/pull/2320
    - Case 3: https://github.com/Expensify/App/pull/22984
- The offending PR has been commented on:
    - Case 1: https://github.com/Expensify/App/pull/15201#discussion_r1333198896
    - Case 2: https://github.com/Expensify/App/pull/2320#discussion_r1333199203
    - Case 3: https://github.com/Expensify/App/pull/22984#discussion_r1333199524
- A discussion in #expensify-bugs has been started: I don't think this is needed. Those bugs could have been caught while testing.
- Determine if we should create a regression test for this bug: I think it may be an overkill to add a regression tests for such cases. So nothing to add here.

**Comment by thienlnam** [comment]
@kavimuru Do you have any additional details on being able to reproduce this one?

**Comment by techievivek** [comment]
Gentle bump @kavimuru ^

**Comment by s-alves10** [comment]
@thienlnam @techievivek 

Can I have a status update on this issue?

**Comment by s-alves10** [comment]
@thienlnam @techievivek 

Can I have a status update on this issue?

**Comment by techievivek** [comment]
We can move forward with the payment here. CC @garrettmknight can we pay the contributors here since the regression period is over, thanks.

**Comment by s77rt** [comment]
Should the payment here be $1000 because this issue is coming from https://github.com/Expensify/App/issues/17531 where we had a regression there

**Comment by garrettmknight** [comment]
Good catch @s77rt - to recap, we didn't pay out the previous issue because we rolled the PR back, opened up this issue with a $2k start price and since this was created as part of a regression, we'll pay 50%.

Summmary of payments for this issue:
- urgency bonus: n/a
- regression penalty: yes
- Reporter: N/A - original issue was reported by Applause
- Contributor: @s-alves10 paid $1000 via Upwork
- Reviewer: @s77rt paid $1000 via Upwork

**Comment by s-alves10** [comment]
@garrettmknight 

I'm not sure why this should be considered as a simple regression of https://github.com/Expensify/App/issues/17531
This solved new issues like https://expensify.slack.com/archives/C049HHMV9SM/p1692747837798019

I hope you explain why
Thanks

**Comment by garrettmknight** [comment]
I think you can consider this like an extension of https://github.com/Expensify/App/issues/17531. If the automation had allowed, we would have reopened that issue, had you fix the regression and paid out at 50% for the regression. If that had happened and we'd separately focused this issue on the other cases (and no regressions were reported) we'd have paid out $1000 as standard at the time. With that in mind, I could see paying $1500 total out for this issue.

**Comment by garrettmknight** [comment]
Added $500 to both contracts.
